pkg/storage/unified/resource/storage_backend_test.go

```diff
@@ -0,0 +1,1030 @@
+			case resourcepb.WatchEvent_DELETED:
+				expectedAction = DataActionDeleted
+			default:
+				t.Fatalf("unexpected event type: %v", tt.eventType)
+			}
+
+			dataKey := DataKey{
+				Namespace:       "default",
+				Group:           "apps",
+				Resource:        "resource",
+				Name:            "test-resource",
+				ResourceVersion: rv,
+				Action:          expectedAction,
+			}
+
+			dataReader, err := backend.dataStore.Get(ctx, dataKey)
+			require.NoError(t, err)
+			dataValue, err := io.ReadAll(dataReader)
+			require.NoError(t, err)
+			assert.Equal(t, objectToJSONBytes(t, testObj), dataValue)
+
+			// Verify metadata was written to metaStore
+			metaKey := MetaDataKey{
+				Namespace:       "default",
+				Group:           "apps",
+				Resource:        "resource",
+				Name:            "test-resource",
+				ResourceVersion: rv,
+				Action:          expectedAction,
+				Folder:          "",
+			}
+
+			m, err := backend.metaStore.Get(ctx, metaKey)
+			require.NoError(t, err)
+			require.NotNil(t, m)
+			require.Equal(t, "test-resource", m.Key.Name)
+			require.Equal(t, "default", m.Key.Namespace)
+			require.Equal(t, "apps", m.Key.Group)
+			require.Equal(t, "resource", m.Key.Resource)
+
+			// Verify event was written to eventStore
+			eventKey := EventKey{
+				Namespace:       "default",
+				Group:           "apps",
+				Resource:        "resource",
+				Name:            "test-resource",
+				ResourceVersion: rv,
+			}
+
+			_, err = backend.eventStore.Get(ctx, eventKey)
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestKvStorageBackend_WriteEvent_ResourceAlreadyExists(t *testing.T) {
+	backend := setupTestStorageBackend(t)
+	ctx := context.Background()
+
+	// Create a test resource first
+	testObj, err := createTestObject()
+	require.NoError(t, err)
+
+	metaAccessor, err := utils.MetaAccessor(testObj)
+	require.NoError(t, err)
+
+	writeEvent := WriteEvent{
+		Type: resourcepb.WatchEvent_ADDED,
+		Key: &resourcepb.ResourceKey{
+			Namespace: "default",
+			Group:     "apps",
+			Resource:  "deployments",
+			Name:      "test-deployment",
+		},
+		Value:      objectToJSONBytes(t, testObj),
+		Object:     metaAccessor,
+		PreviousRV: 0,
+	}
+
+	// First create should succeed
+	rv1, err := backend.WriteEvent(ctx, writeEvent)
+	require.NoError(t, err)
+	require.Greater(t, rv1, int64(0))
+
+	// Try to create the same resource again - should fail with ErrResourceAlreadyExists
+	writeEvent.PreviousRV = 0 // Reset previous RV to simulate a fresh create attempt
+	rv2, err := backend.WriteEvent(ctx, writeEvent)
+	require.Error(t, err)
+	require.Equal(t, int64(0), rv2)
+	require.ErrorIs(t, err, ErrResourceAlreadyExists)
+
+	// Verify the error is the correct type
+	require.Contains(t, err.Error(), "the resource already exists")
+}
+
+func TestKvStorageBackend_ReadResource_Success(t *testing.T) {
+	backend := setupTestStorageBackend(t)
+	ctx := context.Background()
+
+	// First, write a resource to read
+	testObj, err := createTestObject()
```

Comment:
Can we have `createAndWriteTestObject(t, backend)` to avoid duplicating this code in tests?

Or perhaps `writeObject(t, backend, testObj)`, that will fill `WriteEvent` based on the `testObj`?